sympy/core/tests/test_sympify.py

```diff
@@ -647,6 +647,53 @@ def test_sympify_rational_numbers_set():
     assert sympify({'.3', '.2'}, rational=True) == FiniteSet(*ans)
 
 
+def test_sympify_mro():
+    """Tests the resolution order for classes that implement _sympy_"""
+    class a:
+        def _sympy_(self):
+            return Integer(1)
+    class b(a):
+        def _sympy_(self):
+            return Integer(2)
+    class c(a):
+        pass
+
+    assert sympify(a()) == Integer(1)
+    assert sympify(b()) == Integer(2)
+    assert sympify(c()) == Integer(1)
+
+def test_sympify_converter():
+    """Tests the resolution order for classes in converter"""
+    class a:
+        pass
+    class b(a):
+        pass
+    class c(a):
+        pass
+
+    converter[a] = lambda x: Integer(1)
+    converter[b] = lambda x: Integer(2)
+
+    assert sympify(a()) == Integer(1)
+    assert sympify(b()) == Integer(2)
+    assert sympify(c()) == Integer(1)
+
+    class MyInteger(Integer):
+        pass
+
+    if int in converter:
+        int_converter = converter[int]
+    else:
+        int_converter = None
+    converter[int] = MyInteger
+
+    assert sympify(1) == MyInteger(1)
+
+    if int_converter is None:
+        del converter[int]
+    else:
+        converter[int] = int_converter
```

Comment:
This should use `try/finally`. Also it would be better to set a converter for a locally defined temporary class rather than messing with `int`.

Reply:
There are tests for temp defined classes (`a b c`), this is to see if overriding a sympy defined converter works.

I'll change it to a `try` block. When writing this, I thought it might be convenient if SymPy provides a context manager for this, so that users can temporarily change the behaviour of `sympify`/`converter`.

Reply:
I don't think that we should encourage users to do this. Changing the converter for something like `int` would break everything:
```python
In [1]: x + 1
Out[1]: x + 1

In [2]: from sympy.core.sympify import converter

In [3]: converter[int] = lambda e: E

In [4]: x + 1
Out[4]: x + ℯ
```

Reply:
While I agree that it would be a problem if a user would replace the converter with something that doesn't work, I do think this should be tested for (or explicitly forbidden), considering the changes proposed in https://github.com/sympy/sympy/pull/22785 where a future change might introduce a situation where `converter[int] =` does not actually change the behaviour of `sympify` at all (e.g. if it first checks `_sympy_converter`) eventhough the user explicitly indicated to do something different from the SymPy setting.

Reply:
Adding a test is fine but we shouldn't add utilities like a context manager to encourage anyone to do this.

Reply:
Is this resolved now?

Reply:
Yes, looks fine. It seems like overkill to check whether int converter is set or not: it should always be set but not a big deal. Main thing is that even if the test fails the converter will be reset.

Reply:
In https://github.com/sympy/sympy/pull/22785 it would fail, since `converter` would no longer contain SymPy defined converters

Reply:
Oh, I see what you mean! Ok, this is fine.